zcash_primitives/src/merkle_tree.rs

```diff
@@ -110,6 +118,31 @@ impl<Node: Hashable> CommitmentTree<Node> {
             && self.parents.len() == depth - 1
             && self.parents.iter().all(|p| p.is_some())
     }
+}
+
+impl<Node: Hashable> CommitmentTree<Node> {
+    /// Reads a `CommitmentTree` from its serialized form.
+    #[allow(clippy::redundant_closure)]
+    pub fn read<R: Read>(mut reader: R) -> io::Result<Self> {
+        let left = Optional::read(&mut reader, |r| Node::read(r))?;
+        let right = Optional::read(&mut reader, |r| Node::read(r))?;
+        let parents = Vector::read(&mut reader, |r| Optional::read(r, |r| Node::read(r)))?;
```

Comment:
```suggestion
        let left = Optional::read(&mut reader, Node::read)?;
        let right = Optional::read(&mut reader, Node::read)?;
        let parents = Vector::read(&mut reader, |r| Optional::read(r, Node::read))?;
```

Reply:
Rust fails at closure elision here due to a lifetime mismatch. This is pretty common in these `read` scenarios, IME.

Reply:
In fact, if you look just above I have to add `#[allow(clippy::redundant_closure)]` because of this.

Reply:
I checked this again locally on the current PR branch, and it definitely works.